### PR TITLE
S3: Support Multi-region access points

### PR DESCRIPTION
**Pull request by alextwoods**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

1. To make sure we include your contribution in the release notes, please make sure to add description entry for your changes in the "unreleased changes" section of the `CHANGELOG.md` file (at corresponding gem). For the description entry, please make sure it lives in one line and starts with `Feature` or `Issue` in the correct format.

2. For generated code changes, please checkout below instructions first:
  https://github.com/aws/aws-sdk-ruby/blob/version-3/CONTRIBUTING.md

Thank you for your contribution!

### CODE REVIEW

File: gems/aws-sigv4/lib/aws-sigv4/signer.rb [hunk 1]
```diff
@@ -129,6 +129,12 @@ def initialize(options = {})
         [:uri_escape_path, :apply_checksum_header].each do |opt|
           instance_variable_set("@#{opt}", options.key?(opt) ? !!options[opt] : true)
         end
+
+        if options[:signing_algorithm] == :sigv4a
+          raise ArgumentError, 'You are attempting to sign a
```
Comment (mullermp) [inline gems/aws-sigv4/lib/aws-sigv4/signer.rb]: String is multi line.. use `\`

File: services.json [hunk 2]
```diff
@@ -746,7 +746,7 @@
     "models": "s3/2006-03-01",
     "dependencies": {
       "aws-sdk-kms": "~> 1",
-      "aws-sigv4": "~> 1.1"
+      "aws-sigv4": "~> 1.4"
```
Comment (mullermp) [inline services.json:749]: Discussed offline, I think we need to bump this generally for all services.

File: gems/aws-sdk-s3/spec/presigner_spec.rb [hunk 3]
```diff
@@ -329,6 +329,48 @@ module S3
...
+          let(:client) do
+            Aws::S3::Client.new(
+              stub_responses: true,
+              s3_disable_multiregion_access_points: true
+            )
+          end
+
+          it 'raises an ArgumentError' do
+            arn = 'arn:aws:s3::123456789012:accesspoint:mfzwi23gnjvgw.mrap'
+            expect {
```
Comment (mullermp) [inline gems/aws-sdk-s3/spec/presigner_spec.rb]: `do..end` for multiline blocks

File: gems/aws-sdk-s3/spec/presigner_spec.rb [hunk 4]
```diff
@@ -329,6 +329,48 @@ module S3
...
+      context 'MRAP ARNs' do
+        let(:signer) { double('sigv4a_signer') }
+        let(:arn) { 'arn:aws:s3::123456789012:accesspoint:mfzwi23gnjvgw.mrap' }
+
+        it 'creates a presigned url with sigv4a' do
+          # Signer is created twice, return a normal one first
+          expect(Aws::Sigv4::Signer).to receive(:new).and_call_original
+          expect(Aws::Sigv4::Signer)
+            .to receive(:new)
+            .with(hash_including(
```
Comment (mullermp) [inline gems/aws-sdk-s3/spec/presigner_spec.rb]: formatting (in entire test)

File: gems/aws-sdk-s3/lib/aws-sdk-s3/presigner.rb [hunk 5]
```diff
@@ -231,17 +231,23 @@ def sign_but_dont_send(
...
           http_req.endpoint.query = query.join('&') unless query.empty?
 
+          signing_algorithm = :sigv4
+
           # If it's an ARN, get the resolved region and service
           if (arn = context.metadata[:s3_arn])
             region = arn[:resolved_region]
             service = arn[:arn].service
+            region = arn[:arn].is_a?(MultiRegionAccessPointARN) ? '*': arn[:resolved_region]
+            signing_algorithm = arn[:arn].is_a?(MultiRegionAccessPointARN) ? :sigv4a : :sigv4
```
Comment (mullermp) [inline gems/aws-sdk-s3/lib/aws-sdk-s3/presigner.rb:241]: Sigv4 is already the default here

File: gems/aws-sdk-s3/lib/aws-sdk-s3/plugins/arn.rb [hunk 6]
```diff
@@ -113,8 +126,14 @@ def validate_config!(context, arn)
...
-                    'Cannot provide an Outpost Access Point ARN when '\
-                    '`:use_dualstack_endpoint` is set to true.'
+                    'Cannot provide an Outpost Access Point or Multi-region Access Point ARN'\
+                    ' when `:use_dualstack_endpoint` is set to true.'
+            end
+
+            if arn.region.empty? && context.config.s3_disable_multiregion_access_points
+              raise ArgumentError,
+                    'Cannot provide a Multi-region Access Point ARN with '\
+                    ':s3_disable_multiregion_access_points set to true'
```
Comment (mullermp) [inline gems/aws-sdk-s3/lib/aws-sdk-s3/plugins/arn.rb]: Wrap config in backticks

**Review by mullermp — APPROVED** [review]
LGTM! Nice work! I think S3 changelog needs an entry too.

File: services.json (hunk 2 above)
Comment (alextwoods) [reply services.json:749]: Discussed offline - we will leave these as 1.1 for all non s3 gems.

File: gems/aws-sdk-s3/lib/aws-sdk-s3/presigner.rb (hunk 5 above)
Comment (alextwoods) [reply gems/aws-sdk-s3/lib/aws-sdk-s3/presigner.rb:241]: Wanted this line to match the above - I think its easier to read here than:
```
 signing_algorithm = : sigv4a if arn[:arn].is_a?(MultiRegionAccessPointARN) 
```

**Pull request merged**